apps/sms/js/activity_handler.js

```diff
@@ -125,21 +125,26 @@ var ActivityHandler = {
       threadId: null
     };
 
-    // try to get a thread from number
-    // if no thread, promise is rejected and we try to find a contact
-    return MessageManager.findThreadFromNumber(viewInfo.number).then(
-      function onResolve(threadId) {
-        viewInfo.threadId = threadId;
-      },
-      function onReject() {
-        return ActivityHandler._findContactByNumber(viewInfo.number)
-          .then( (contact) => viewInfo.contact = contact);
-      }
-    )
-    // case no contact and no thread id: gobble the error
-    .catch(() => {})
-    // finally call toView whatever contact and threadId we have.
-    .then( () => this.toView(viewInfo) );
+    var threadPromise;
+    if (viewInfo.number) {
+      // try to get a thread from number
+      // if no thread, promise is rejected and we try to find a contact
+      threadPromise = MessageManager.findThreadFromNumber(viewInfo.number).then(
+        function onResolve(threadId) {
+          viewInfo.threadId = threadId;
+        },
+        function onReject() {
+          return ActivityHandler._findContactByNumber(viewInfo.number)
+            .then((contact) => viewInfo.contact = contact);
+        }
+      )
+      // case no contact and no thread id: gobble the error
+      .catch(() => {});
+    }
+
+    return (threadPromise || Promise.resolve()).then(
+      () => this.toView(viewInfo, !!viewInfo.threadId /* focusComposer */)
```

Comment:
for the "focusComposer" boolean, it's quite misleading, because we'll focus also in the case of a "new" that enters the composer (so no threadId); but that's handled by ThreadUI.

I'd say we want to focus if we have a number. And we could even reuse the information in ThreadUI.handleActivity instead of having the same "if".

So I'd do a "var focusComposer = false" before the if, and set it to true inside the if. (or something similar, up to you). And then use it here. And clean up ThreadUI.handleActivity :)